### DISCUSSION TITLE
UEFI lesson to get Current Link Speed for PCI devices.

### DISCUSSION THREAD

**Discussion by ayushdas2000** [Kostr/UEFI-Lessons#4] [General]
Hi Kostr. Been Learning Uefi development from these lessons. I am trying to write a uefi application to get Current Link speed for pci devices right now but havent been able to reach a working code. Would appreciate any leads or a lesson for the same. Thanks

**Comment by Kostr** [comment]
Hi @ayushdas2000. There is a lesson about reading PCIe configuration space [Lesson_30](https://github.com/Kostr/UEFI-Lessons/tree/master/Lessons/Lesson_30). You just need to understand how to find current link speed from the PCIe configuration space. 
Basically you need to search for the necessary capability starting from the `Capability Pointer` register. And then read `Current Link speed` from that capability. Consult the PCIe specification or lspci sources for the implementation details.

You can also grab the configuration space parsing logic from my [pcie-mdr](https://github.com/Kostr/pcie-mdr/blob/32381a52e046151c842764163e4b2e1412a96f39/src/pcie.cpp#L111) application developed for the openbmc.

**Reply by ayushdas2000** [reply]
Thanks a lot. Will try this

**Reply by ayushdas2000** [reply]
Was able to get this done. Thanks a lot
